File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
    stack<long> st;
    long mini;
public:
    MinStack() {
        mini = INT_MAX;
    }

    void push(int val) {
        long value = val;
        if (st.empty()) {
            mini = value;
            st.push(value);
        } else {
            if(val<mini){
                st.push(2*value-mini);
                mini = value;
            }
            else{
                st.push(value);
            }
        }
    }

    void pop() {
        if (st.empty()) return;
        long el = st.top();
        st.pop();
        if(el<mini){
            mini = 2*mini-el;
        }
    }

    int top() {
        if (st.empty()) return -1;
        long el = st.top();
        if(el<mini) return mini;
        return el;
    }

    int getMin() {
        if (!st.empty()) {
            return mini;
        }
        return -1;
    }
};

File: 0155-min-stack/0155-min-stack.java
class MinStack {
    Stack<Long> st;
    long minVal;
    public MinStack() {
        st = new Stack<>();
    }
    
    public void push(int val) {
        long value = val;
        if(st.isEmpty()){
            minVal = value;
            st.push(value);
        }else{
            if(value<minVal){
                st.push((2*value-minVal));
                minVal = value;
            }else st.push(value);
        }
    }
    
    public void pop() {
        long el = st.pop();
        if(el<minVal){
            minVal = 2*minVal-el;
        }
    }
    
    public int top() {
        long el = st.peek();
        if(el<minVal) return (int)minVal;
        return (int)el;
    }
    
    public int getMin() {
        return (int)minVal;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */